### PR TITLE
Add functionality to remove values from inventory and feature profiles

### LINKED ISSUE #300: Delete values from `features_listed_values` and feature profiles
Also related:

- [ ] #319

### PR DESCRIPTION
**Pull request by MariAnShum**
closes #300

### CODE REVIEW

**Comment by MariAnShum** [comment]
So far I am working with removing from the inventory only.

Ho should removed values be treated in feature profiles? Should they become custom or not_stated? To my mind, custom values would sound more logical and suitable for possible value movement. Perhaps both options should be availible?

File: langworld_db_data/removers/listed_value_remover.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+
+class ListedValueRemover(Remover):
+    def remove_listed_value(
+        self,
+        id_of_value_to_remove: str,
+    ) -> dict[str, str]:
+        removed_value_information = self._remove_from_inventory_of_listed_values(
+            id_of_value_to_remove
+        )
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:17]: please add docstring to the method that clarifies what is returned
`removed_value_information` is too broad a variable name. It's not clear what the variable contains

File: langworld_db_data/removers/listed_value_remover.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                f"Perhaps it is invalid or does not exist"
+            )
+        else:
+            return removed_value_information
+
+    def _remove_from_inventory_of_listed_values(
+        self,
+        id_of_value_to_remove: str,
+    ) -> dict[str, str]:
+        removed_value_information = {}
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:32]: same as above

File: langworld_db_data/removers/listed_value_remover.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        line_number_of_value_to_remove = 0
+
+        for i, row in enumerate(rows):
+            if row["id"] == id_of_value_to_remove:
+                line_number_of_value_to_remove = i
+                removed_value_information["feature_id"] = row["feature_id"]
+                removed_value_information["en"] = row["en"]
+                removed_value_information["ru"] = row["ru"]
+                break
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:44]:
```suggestion
            if row["id"] != id_of_value_to_remove:
                continue
            line_number_of_value_to_remove = i
            removed_value_information = {
                key: row[key] for key in ("feature_id", "en", "ru")
            }
```

File: langworld_db_data/removers/listed_value_remover.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        rows_with_updated_value_indices = self._update_value_indices_in_inventory(
+            rows=rows,
+            id_of_value_to_remove=id_of_value_to_remove,
+        )
+
+        # If value not found, inventory will remain intact
+        rows_without_removed_value = (
+            rows_with_updated_value_indices[:line_number_of_value_to_remove]
+            + rows_with_updated_value_indices[line_number_of_value_to_remove + 1 :]
+        )
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:55]: this looks a bit strange: looks like `self._update_value_indices_in_inventory()` still returns, among others, the row that should be removed. But an argument that this function has is called `id_of_value_to_remove`. So it gets something to remove, but doesn't remove it? Because otherwise you wouldn't need the next statement.

It makes sense that you want to separate these two steps (removing a row and updating IDs). But maybe if you run these operations in a reverse order, it would make more sense and be more logical (cutting a row, then renaming IDs sounds obvious to a reader)

File: langworld_db_data/removers/listed_value_remover.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    def _remove_from_feature_profiles(self):
+        self._update_value_ids_in_feature_profiles()
+
+    @staticmethod
+    def _update_value_indices_in_inventory(
+        rows: list[dict[str, str]],
+        id_of_value_to_remove: str,
+    ) -> list[dict[str, str]]:
+        feature_id = "-".join(id_of_value_to_remove.split(ID_SEPARATOR)[:2])
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:74]: I'm starting to think that we might want to create universal helper methods for getting feature ID or value index from value ID. We're using this `split[blah-blah]` thing way too much. Maybe a TODO for a future refactoring PR.

File: langworld_db_data/removers/listed_value_remover.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        rows_with_updated_indices = rows
+
+        for i, row in enumerate(rows):
+            if row["feature_id"] != feature_id:
+                continue
+            current_value_index = int(row["id"].split("-")[2])
+            index_of_value_to_remove = int(id_of_value_to_remove.split("-")[2])
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:81]: that's what I'm talking about... Could you create an issue and we'll take care of it in due course?

File: tests/test_removers_listed_value_remover.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    OUTPUT_DIR_FOR_LISTED_VALUE_REMOVER_FEATURE_PROFILES,
+)
+
+GS_FILE_WITH_LISTED_VALUES_REMOVING_THE_LAST = (
+    DIR_WITH_REMOVERS_TEST_FILES / "features_listed_values_gold_standard_for_removing_the_last.csv"
+)
+GS_FILE_WITH_LISTED_VALUES_REMOVING_IN_MIDDLE = (
+    DIR_WITH_REMOVERS_TEST_FILES
+    / "features_listed_values_gold_standard_for_removing_in_middle.csv"
+)
```
Comment (lemontree210) [inline tests/test_removers_listed_value_remover.py:21]: how about `_the_first`? It could be the same as `_in_middle`, but it's nevertheless expected that we cover both edge cases (first and last)

File: tests/test_removers_listed_value_remover.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    return ListedValueRemover(
+        input_file_with_listed_values=INPUT_FILE_WITH_LISTED_VALUES_FOR_REMOVERS,
+        output_file_with_listed_values=DIR_WITH_REMOVERS_TEST_FILES
+        / "features_listed_values_output_value_remover.csv",
+        input_dir_with_feature_profiles=DIR_WITH_REMOVERS_FEATURE_PROFILES,
+        output_dir_with_feature_profiles=OUTPUT_DIR_FOR_LISTED_VALUE_REMOVER_FEATURE_PROFILES,
+    )
+
+
+# Tests
```
Comment (lemontree210) [inline tests/test_removers_listed_value_remover.py:41]:
```suggestion
```

File: tests/test_removers_listed_value_remover.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    for bad_value_id in ["A-5-256", "S-256", "ABC"]:
+        with pytest.raises(ListedValueRemoverError, match="not found. Perhaps"):
+            test_remover.remove_listed_value(bad_value_id)
+
+
+# What should the module do if it is asked to pop a value which does not exist?
+
+
+# How should the removed values be changed in profiles? Should they become custom or not_stated?
+# To me, it would be more logical to make them custom, again for the sake of possible movement
```
Comment (lemontree210) [inline tests/test_removers_listed_value_remover.py:90]: I agree with `custom`. Maybe the program should also alert the user about it. Maybe save a text file with names of profiles, in which the value was found and got the type `custom`.

File: tests/test_removers_listed_value_remover.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+# However, perhaps both options may be necessary to provide
+
+# Should we also be able to remove custom values?
```
Comment (lemontree210) [inline tests/test_removers_listed_value_remover.py:93]: I don't think that we need functionality to automatically remove custom values from profiles

**Review by lemontree210 — CHANGES_REQUESTED** [review]

File: tests/test_removers_listed_value_remover.py (hunk 7 above)
Comment (MariAnShum) [reply tests/test_removers_listed_value_remover.py]: Done!

File: langworld_db_data/removers/listed_value_remover.py [hunk 11]
```diff
@@ -15,7 +15,13 @@ def remove_listed_value(
         removed_value_information = self._remove_from_inventory_of_listed_values(
             id_of_value_to_remove
         )
-        self._remove_from_feature_profiles()
+        """
+        Contains the feature ID and English and Russian names of the removed value.
+        """
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:20]: a dictionary is best named `value_for_key` or `key_to_value`. `_information` is not informative :) The docstring you added doesn't make it clear what the keys and values are either.

File: langworld_db_data/removers/listed_value_remover.py [hunk 12]
```diff
@@ -15,7 +15,13 @@ def remove_listed_value(
...
+        self._remove_from_feature_profiles(
+            id_of_value_to_remove=id_of_value_to_remove,
+            english_name_of_value_to_remove=removed_value_information["en"],
+        )
 
         if not removed_value_information:
             raise ListedValueRemoverError(
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:27]: fix codecov issue?

File: langworld_db_data/removers/listed_value_remover.py [hunk 13]
```diff
@@ -30,55 +36,104 @@ def _remove_from_inventory_of_listed_values(
...
+            if row["id"] != id_of_value_to_remove:
+                continue
+            line_number_of_value_to_remove = i
+            removed_value_information = {key: row[key] for key in ("feature_id", "en", "ru")}
 
         # If value not found, inventory will remain intact
         rows_without_removed_value = (
-            rows_with_updated_value_indices[:line_number_of_value_to_remove]
-            + rows_with_updated_value_indices[line_number_of_value_to_remove + 1 :]
+            rows[:line_number_of_value_to_remove] + rows[line_number_of_value_to_remove + 1 :]
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:55]: BTW can we just use 'list.remove()` method here and drop the `line_number_to_remove` and `enumerate()` altogether?

File: langworld_db_data/removers/listed_value_remover.py [hunk 14]
```diff
@@ -30,55 +36,104 @@ def _remove_from_inventory_of_listed_values(
...
+                )
+                with open(file_with_changed_profiles_report, "a", encoding="utf-8") as report:
+                    print(f"Adding '{file.stem}' to the changed files report")
+                    report.write(f"{file.stem}\n")
+            else:
+                print(f"{file.stem} is not changed")
 
-    def _remove_from_feature_profiles(self):
-        self._update_value_ids_in_feature_profiles()
+        with open(file_with_changed_profiles_report, "a", encoding="utf-8") as report:
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:121]: as far as I remember, you don't need to specity utf-8 because it's the default

File: langworld_db_data/removers/listed_value_remover.py [hunk 15]
```diff
@@ -30,55 +36,104 @@ def _remove_from_inventory_of_listed_values(
...
+        else:
+            return removed_value_information
+
+    def _remove_from_feature_profiles(
+        self,
+        id_of_value_to_remove: str,
+        english_name_of_value_to_remove: str,
+    ):
+        if not self.output_dir_with_feature_profiles.exists():
+            self.output_dir_with_feature_profiles.mkdir()
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:86]: that's a strange place for these statements, I think. It would be clearer to do these things at initialization of object

File: langworld_db_data/removers/listed_value_remover.py [hunk 16]
```diff
@@ -30,55 +36,104 @@ def _remove_from_inventory_of_listed_values(
...
+
+        english_name_of_value_to_remove = english_name_of_value_to_remove.replace(",", "")
+        english_name_of_value_to_remove = english_name_of_value_to_remove.replace(" ", "_")
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:89]:
```suggestion
        english_name_of_value_to_remove = english_name_of_value_to_remove.replace(",", "").replace(" ", "_")
```

Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py:89]: I also think that a comment is needed for this block of statements, saying that you're creating a report.

BTW I don't think we created any reports for adding listed values, so why here?
Maybe logging could be done in a separate PR, with use of `logging` module.

File: tests/test_removers_listed_value_remover.py [hunk 17]
```diff
@@ -38,9 +53,32 @@ def test_remover():
...
+        output_file=test_remover.output_file_with_listed_values,
+        gold_standard_file=GS_FILE_WITH_LISTED_VALUES_REMOVING_A_9_1,
+    )
+
+    (test_remover.output_dir_with_feature_profiles / "corsican.csv").unlink()
+    (test_remover.output_dir_with_feature_profiles / "ukrainian.csv").unlink()
+    (
+        test_remover.output_dir_with_feature_profiles
+        / "changed_profiles_A-9-1_No_diphthongs_and_triphthongs.txt"
+    ).unlink()
```
Comment (lemontree210) [inline tests/test_removers_listed_value_remover.py:75]: is it really necessary? `check_existence...` already unlinks files.

I don't think you need to unlink the output directory. Anyway, you have to test both cases in some separate test function: when the output directory doesn't exist and when it already exists. 

Since I suggested to move creation of non-existent output dir to `__init__()` method of `ListedValueRemover`, a test can be run for that method. And here, it's OK to assume that the output dir already exists.

Not sure I made myself clear here :)

**Review by lemontree210 — CHANGES_REQUESTED** [review]

File: langworld_db_data/removers/listed_value_remover.py (hunk 13 above)
Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py:58]: To use `remove()`, we will have to alter `removed_value_information` formulation by including the key "id" with the value's ID as its value. This will entail alternation of some tests and, perhaps, some other elements in the code of the class `ListedValueRemover`. If this is alright, I will apply these changes

File: langworld_db_data/removers/listed_value_remover.py (hunk 15 above)
Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py:89]: I have made an attempt to put these lines into `__init__()`, calling `super.__init__()` before them, but then an error appears in every test for this module. This error seems unfamiliar to me

File: langworld_db_data/removers/listed_value_remover.py (hunk 16 above)
Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py]: So far I commented out all the pieces of code that have to do with logging in TXT format. Perhaps they might be useful for logging with the help of the module.

File: tests/test_removers_listed_value_remover.py (hunk 17 above)
Comment (MariAnShum) [reply tests/test_removers_listed_value_remover.py]: > is it really necessary? `check_existence...` already unlinks files.

At the time when I worked on this method, for an unknown reason these files persistently remained after applying tests. Now, I have tried to comment these lines out and there were no extra files left after the tests, which is quite strange to me (but I erased these lines anyway).

> Anyway, you have to test both cases in some separate test function: when the output directory doesn't exist and when it already exists.

But, if I am not mistaken, in actual practice with the DB the input directory is always the output directory too so the output directory always exists and checking its existence is needed for tests only. Should I write the test anyway?

> Since I suggested to move creation of non-existent output dir to `__init__()` method of `ListedValueRemover`

Introducing `__init__` (with calling `super.__init__()` inside) causes an error in every test:
```
tests/test_removers_listed_value_remover.py::test__remove_from_feature_profiles - TypeError: ListedValueRemover.__init__() got an unexpected keyword argument 'input_file_with_listed_values'
```
How could it possibly be fixed?

**Review by lemontree210 — COMMENTED** [review]
do you plan to remove commented lines before merging into `master`?

**Comment by MariAnShum** [comment]
> do you plan to remove commented lines before merging into `master`?

Yes, I have removed all the commented codelines (including those related to reports in the `txt` format).

File: langworld_db_data/removers/listed_value_remover.py (hunk 12 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py:30]: `codecov` is still complaining about `raise ListedValueRemoverError` in line 30 not being covered by tests

**Review by lemontree210 — CHANGES_REQUESTED** [review]

File: langworld_db_data/removers/listed_value_remover.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+    ) -> dict[str, str]:
+        removed_value_information = self._remove_from_inventory_of_listed_values(
+            id_of_value_to_remove
+        )
+        """
+        The dictionary contains three items:
+        - key: "feature_id", value: the feature ID of the removed value,
+        - key: "en", value: its English name,
+        - key: "ru", value: its Russian name.
+        """
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:23]:
```suggestion
        removed_value = self._remove_from_inventory_of_listed_values(
            id_of_value_to_remove
        )
        """
        The dictionary contains three items:
        - key: "feature_id", value: the feature ID of the removed value,
        - key: "en", value: its English name,
        - key: "ru", value: its Russian name.
        """
```
Let's just remove `_information` from variable name then.

Also: I remembered that we now have two additional columns `description_formatted_en` and `description_formatted_ru`.

So, my suggestion is: let's make the keys in this dictionary identical to columns in `features_listed_values.csv`,
adding `id`, `description_formatted_en`, `description_formatted_ru`. And we can then write in the docstring that
the keys are identical to column names in inventory of listed values.

`id` may be excessive but very easy to fill anyway. Two other columns are actually necessary because in the future we might find ourselves removing (for future pasting) a value that has an HTML description

**Review by lemontree210 — CHANGES_REQUESTED** [review]

File: langworld_db_data/removers/listed_value_remover.py (hunk 13 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py:58]: See https://github.com/Jazyki-Mira/langworld_db_data/pull/301#discussion_r1863045405

File: langworld_db_data/removers/listed_value_remover.py (hunk 11 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py]: See https://github.com/Jazyki-Mira/langworld_db_data/pull/301#discussion_r1863045405

File: langworld_db_data/removers/listed_value_remover.py (hunk 18 above)
Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py]: Done!

**Comment by MariAnShum** [comment]
> `codecov` is still complaining about raise `ListedValueRemoverError` in line 30 not being covered by tests

I have put the very same check into the method where `removed_value` is received initially --  `_remove_from_inventory_of_listed_values`. Could this check possibly be deleted from `remove_listed_value`?

File: tests/test_removers_listed_value_remover.py (hunk 17 above)
Comment (MariAnShum) [reply tests/test_removers_listed_value_remover.py]: I have also thought once again of introducing `remove()` to this method (https://github.com/Jazyki-Mira/langworld_db_data/pull/301#discussion_r1841732054) but could not come up with any ways to do it. However I found a way to make the `enumerate()` cycle shorter

File: langworld_db_data/removers/listed_value_remover.py (hunk 18 above)
Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py]: I have also thought once again of introducing `remove()` to this method (https://github.com/Jazyki-Mira/langworld_db_data/pull/301#discussion_r1841732054) but could not come up with any ways to do it. However I found a way to make the `enumerate()` cycle shorter

Comment (MariAnShum) [reply langworld_db_data/removers/listed_value_remover.py]: If the inventory now has a slightly changed format, I think this change should also be applied to methods and tests for `ListedValueAdder`. Should I create an issue for this?

File: langworld_db_data/removers/listed_value_remover.py [hunk 19]
```diff
@@ -12,30 +12,29 @@ def remove_listed_value(
         self,
         id_of_value_to_remove: str,
     ) -> dict[str, str]:
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:14]: I think we should add a docstring to the whole method because it's public. It should definitely include a description of the dictionary being returned.

**Review by lemontree210 — COMMENTED** [review]
Thank you for the changes! I'm sorry if we already discussed it and I forgot, but how about fixing codecov issue by creating a test that checks if `ListedValueRemoverError` gets thrown if `removed_value` is empty?

UPD found your https://github.com/Jazyki-Mira/langworld_db_data/pull/301#issuecomment-2519545397

File: langworld_db_data/removers/listed_value_remover.py (hunk 13 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py:54]: OK, let's leave it as it is. Changing a list while looping over it is not a good idea.

File: langworld_db_data/removers/listed_value_remover.py [hunk 20]
```diff
@@ -30,55 +36,104 @@ def _remove_from_inventory_of_listed_values(
...
+                f"Value ID {id_of_value_to_remove} not found. "
+                f"Perhaps it is invalid or does not exist"
+            )
+        else:
+            return removed_value_information
+
+    def _remove_from_feature_profiles(
+        self,
+        id_of_value_to_remove: str,
+        english_name_of_value_to_remove: str,
```
Comment (lemontree210) [inline langworld_db_data/removers/listed_value_remover.py:82]: note that `english_name_of_value_to_remove` is not used in the code of this method, so you can remove it here and above (where you call this method).

File: langworld_db_data/removers/listed_value_remover.py (hunk 15 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py]: well I just put in into `Remover` :)
see 9e7c808

File: langworld_db_data/removers/listed_value_remover.py (hunk 20 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py]: UPD I did it myself in e483bef

File: langworld_db_data/removers/listed_value_remover.py (hunk 18 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py]: > If the inventory now has a slightly changed format, I think this change should also be applied to methods and tests for `ListedValueAdder`. Should I create an issue for this?

Yes, please do 👍🏻

**Comment by lemontree210** [comment]
> > `codecov` is still complaining about raise `ListedValueRemoverError` in line 30 not being covered by tests
> 
> I have put the very same check into the method where `removed_value` is received initially -- `_remove_from_inventory_of_listed_values`. Could this check possibly be deleted from `remove_listed_value`?

Good point. I am removing this check from `remove_listed_value`.

File: langworld_db_data/removers/listed_value_remover.py (hunk 19 above)
Comment (lemontree210) [reply langworld_db_data/removers/listed_value_remover.py:14]: UPD did it myself

File: tests/test_removers_listed_value_remover.py (hunk 17 above)
Comment (lemontree210) [reply tests/test_removers_listed_value_remover.py]: closing this thread as these lines were removed

**Comment by lemontree210** [comment]
I made some changes, you can use [this link](https://github.com/Jazyki-Mira/langworld_db_data/pull/301/files/af80a0508f9b3f442accac5bd8d67a94057ea5cd..9f1ee6fef8328b2ee3de3ee071e11834d648ed05) to look at changes from those 4 commits

please let me know if you're OK with the changes

**Review by lemontree210 — APPROVED** [review]
LGTM

**Comment by MariAnShum** [comment]
> I made some changes, you can use [this link](https://github.com/Jazyki-Mira/langworld_db_data/pull/301/files/af80a0508f9b3f442accac5bd8d67a94057ea5cd..9f1ee6fef8328b2ee3de3ee071e11834d648ed05) to look at changes from those 4 commits
> 
> please let me know if you're OK with the changes

Yes, I am OK with all of the changes listed on that page
